### PR TITLE
Tracks: Add PHP client

### PR DESCRIPTION
**Pull request by psealock**
### All Submissions:

* [x] Have you followed the [WooCommerce Contributing guideline](https://github.com/woocommerce/woocommerce/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Add a PHP Tracks client to expose methods of recording user events via PHP. This PR borrows heavily from the Jetpack version.

### How to test the changes in this Pull Request:

1. Pull this branch down and make sure you've enabled the `woocommerce_allow_tracking` flag.
2. Update a product description (or any other property).
3. Wait 15 minutes and check Tracks Live View for your update product event, https://mc.a8c.com/tracks/live/?eventname=wca_test_update_product&user=&useragent=&rejects=%F0%9F%91%8E

_Note:_ The event is in the `rejects` table because our event name prefix hasn't been whitelisted yet.

### Known issues not addressed in this PR

* Site and user identification.
* Revenue and Product info similar to wc-tracking.

### CODE REVIEW

File: includes/libraries/tracks/client.php
Comment (timmyc) [inline includes/libraries/tracks/client.php]: I think we should also have some logic here, at the very least a check to see if tracker is enabled, but also maybe a filter to allow for disabling of tracks collection all-together.

Comment (psealock) [reply includes/libraries/tracks/client.php]: I updated to this in the renamed file (class-tracks.php).
```php
/**
 * Don't track users who haven't opted-in to tracking or if a filter
 * has been applied to turn it off.
 */
if (
	'yes' !== get_option( 'woocommerce_allow_tracking' ) &&
	! apply_filters( 'woocommerce_apply_user_tracking', true )
) {
	return false;
}
```

File: includes/class-wc-user-tracking.php
Comment (mikejolley) [inline includes/class-wc-user-tracking.php]: Is this a shared library? I think this needs to check if class exists first.

File: includes/libraries/tracks/class-tracks-client.php
Comment (claudiosanches) [inline includes/libraries/tracks/class-tracks-client.php]: As you said, this library is inspirited by Jetpack, but seems modified for WooCommerce at least, since there's options only available on WooCommerce and also filters... So this class name seems generic, probably will cause problems, we probably could move this to another folder and stop look at it as a external library.

File: includes/libraries/tracks/class-tracks-client.php
Comment (claudiosanches) [inline includes/libraries/tracks/class-tracks-client.php]: Each class load another class, this is very easy to create issues...

File: includes/class-wc-user-tracking.php
Comment (claudiosanches) [inline includes/class-wc-user-tracking.php]: This file is called `class-wc-user-tracking.php` but there is no class inside... This probably will be a nice place to load all classes from this library and init it properly.

File: includes/libraries/tracks/class-tracks-client.php
Comment (claudiosanches) [inline includes/libraries/tracks/class-tracks-client.php]: This sound likes `Tracks_Event|array` and is better instead of `mixed`, I don't think that should work with bool, strings and so on.
There's a lot of docblocks in this classes that doesn't describe any param.

File: includes/libraries/tracks/class-tracks-client.php
Comment (claudiosanches) [inline includes/libraries/tracks/class-tracks-client.php]: Why this already comes with deprecated methods?
As I said before, this library seems at least know built for WooCommerce...

File: includes/libraries/tracks/class-tracks-client.php
Comment (claudiosanches) [inline includes/libraries/tracks/class-tracks-client.php]: Are we keeping the `jetpack:` ID?

File: includes/libraries/tracks/class-tracks-event.php
Comment (claudiosanches) [inline includes/libraries/tracks/class-tracks-event.php]: You can just use one `unset()` here.

File: includes/libraries/tracks/class-tracks-client.php
Comment (claudiosanches) [inline includes/libraries/tracks/class-tracks-client.php]: I don't see this method have been used inside this code, but probably nice to check if `Jetpack` exists first here.

**Review by claudiosanches — CHANGES_REQUESTED** [review]
This class sounds like an external library, but since this is so custom to WooCommerce we could move into a new `track` directory. And there is a few methods already coming as deprecated or not used, so probably better to clean up this.

**Comment by psealock** [comment]
Thanks for the reviews @mikejolley @claudiosanches !

Changes made:
* Moved all files under `/includes/tracks/`
* Renamed classes to less generic `WC_Tracks_*`
* Remove deprecated/unused functions
* Remove `mixed` from doc blocks to more specific types
* Renamed `wc-user-tracking.php` to `wc-site-tracking.php`

In regard to identifying sites or users, I'd like to address this separately because more thought will be required in how we want to track in this regard. For the time being, I've left those functions almost as they are found in Jetpack with a few `@TODO`s.

>This probably will be a nice place to load all classes from this library and init it properly.

I am initialising and loading classes in the `wc-site-tracking.php`. I'm not entirely sure if its done properly, but any feedback on that would be great.

**Comment by timmyc** [comment]
@mikejolley / @claudiosanches could we get another review of this PR? We would like to get this merged soon so we can work on adding in usage/collection that utilize the new client. 🙇

**Comment by mikejolley** [comment]
@timmyc @psealock @claudiosanches 

Could you create a new feature branch for this? `feature/tracks`, then target that branch with the PRs instead of master. That way we can review this WIP PR by PR without merging into core. We'll merge into core once feature complete and comms plans are in place.

**Comment by psealock** [comment]
Yup, thanks @mikejolley that makes sense. This PR is aimed at merge into `feature/add-tracks`, which is a fork off master.

**Comment by psealock** [comment]
Unless there are any more comments, questions, or objections coming, I'm going to merge this into the feature branch tomorrow. Feedback is welcome!

File: includes/tracks/class-wc-tracks-client.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks-client.php]: Best to use `wp_safe_remote_get` here?

File: includes/tracks/class-wc-tracks-client.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks-client.php]: r/e redirection and timeout, why specifically these values? Is timeout of 1 enough? If it does timeout, does this retry, and could that lead to performance issues if the host was unreachable for a period of time?

File: includes/tracks/class-wc-tracks-client.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks-client.php]: You can use `wp_remote_retrieve_response_code` here to avoid this logic.

File: includes/tracks/class-wc-tracks-client.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks-client.php]: Is there a difference between this and https://codex.wordpress.org/Function_Reference/current_time

`current_time( 'timestamp', true )`

File: includes/tracks/class-wc-tracks-client.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks-client.php]: Is this being changed in another PR?

File: includes/tracks/class-wc-tracks-client.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks-client.php]: Could use `wc_setcookie` helper here. If those other 2 checks are worth (?) you could improve `wc_setcookie` in the process.

File: includes/tracks/class-wc-tracks-client.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks-client.php]: This probably needs some sort of sanitization before returning it because it's saved straight to meta.

File: includes/tracks/class-wc-tracks-event.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks-event.php]: sanitize? `esc_url` or `esc_url_raw`?

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks.php]: Is this check enough? In other places we have code like this:

```
class_exists( 'Jetpack' ) && is_callable( 'Jetpack::is_active' ) && Jetpack::is_active()
```

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks.php]: Not sure why this is jetpack_ prefixed or if it matters. Just pointing it out.

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks.php]: Without unique (the false 4th prop) wouldn't this make multiple meta entries with the same key?

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks.php]: wc_setcookie

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks.php]: Not sure if here or in other places, but do we need to check the capabiltiies of the user?

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks.php]: Why not check this earlier/first before validating user etc?

File: includes/tracks/wc-site-tracking.php
Comment (mikejolley) [inline includes/tracks/wc-site-tracking.php]: I assume this file is a sample we'll update and review later.

**Review by mikejolley — COMMENTED** [review]
Added some feedback.

File: includes/tracks/class-wc-tracks-client.php
Comment (psealock) [reply includes/tracks/class-wc-tracks-client.php]: Yes indeed. I've left the identity aspects to https://github.com/woocommerce/woocommerce/pull/22676

File: includes/tracks/class-wc-tracks-client.php
Comment (psealock) [reply includes/tracks/class-wc-tracks-client.php]:
```
print_r( current_time( 'timestamp', true ) );        // 1550451155
print_( WC_Tracks_Client::build_timestamp();         // 1550451155739
```
Calypso events in Tracks also use a 13 digit timestamp

File: includes/tracks/class-wc-tracks-client.php
Comment (psealock) [reply includes/tracks/class-wc-tracks-client.php]: Cool, I will make a note of this comment in #22676 and continue the conversation there.

File: includes/tracks/class-wc-tracks-client.php
Comment (psealock) [reply includes/tracks/class-wc-tracks-client.php]: Same as above... will continue this in #22676

File: includes/tracks/class-wc-tracks.php
Comment (psealock) [reply includes/tracks/class-wc-tracks.php]: Thanks! I'm addressing in the next PR

File: includes/tracks/class-wc-tracks.php
Comment (psealock) [reply includes/tracks/class-wc-tracks.php]: Absolutely. In https://github.com/woocommerce/woocommerce/pull/22746 I've added a check so none of the actions are even added if a user is opted out. 

I think it makes sense to leave the check here as well in case this class is used elsewhere.

File: includes/tracks/wc-site-tracking.php
Comment (psealock) [reply includes/tracks/wc-site-tracking.php]: This is an event that we'll be tracking, so its best to review now so that this event doesn't go without some scrutiny.

File: includes/tracks/class-wc-tracks.php
Comment (psealock) [reply includes/tracks/class-wc-tracks.php]: In https://github.com/woocommerce/woocommerce/pull/22676 I change to `update_user_meta` but I'll make a note to ensure multiple entries can't happen.

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [reply includes/tracks/class-wc-tracks.php]: Could be done earlier in this method also since it disables tracking.

File: includes/tracks/wc-site-tracking.php
Comment (mikejolley) [reply includes/tracks/wc-site-tracking.php]: Ok; well in that case:

- This would probably make sense as a class, with a method to add all the hooks and a method per event.
- Would need to know more about intention of the events to see if the used hook is a good fit. For example the current event runs on edit_post which is fine for WP Admin, but won't track product saves outside of admin e.g. edits by plugins

Comment (psealock) [reply includes/tracks/wc-site-tracking.php]: >Would need to know more about intention of the events to see if the used hook is a good fit. For example the current event runs on edit_post which is fine for WP Admin, but won't track product saves outside of admin e.g. edits by plugins.

Great question. I think the nature of these events is to make inferences about usage of the Woo interface only. So modification of products by plugins don't need to be tracked. @warrendholmes is that correct?

File: includes/tracks/class-wc-tracks.php
Comment (psealock) [reply includes/tracks/class-wc-tracks.php]: Are you asking if we should track user role here? If so, that logic would probably go in `WC_Tracks`

File: includes/tracks/class-wc-tracks-client.php
Comment (psealock) [reply includes/tracks/class-wc-tracks-client.php]: I removed `timeout` in favour of the default (5). These values are used by the Jetpack implementation and I can't find any reasoning behind them. The timeout seems short, but I don't know why redirection would be used, so limiting it makes sense.

**Comment by psealock** [comment]
Thanks for the insightful reviews @mikejolley !

I've addressed your comments with changes, follow up questions, or moved relevant comments to the next PR.

I created a class `WC_Site_Tracking` with an init function to add all the actions and a `is_tracking_enabled` function. I'm not certain of the best practices when it comes to instantiating a class like this. Would a singleton make more sense?

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [reply includes/tracks/class-wc-tracks.php]: No, more asking whether tracking should be enable only when an admin/shop manager type user is active, or any user in wp admin.

File: includes/tracks/class-wc-tracks.php
Comment (mikejolley) [inline includes/tracks/class-wc-tracks.php]: Just noticed, this should be OR. If tracking is not yes, OR the filter is used, return.

File: includes/tracks/class-wc-site-tracking.php
Comment (mikejolley) [inline includes/tracks/class-wc-site-tracking.php]: Since this is in a class now you don't need to use woocommerce_ prefixes here which might make the class easier to read :) Not a blocker.

File: includes/tracks/class-wc-site-tracking.php
Comment (mikejolley) [inline includes/tracks/class-wc-site-tracking.php]: should be OR ||

**Comment by mikejolley** [comment]
> I created a class `WC_Site_Tracking` with an init function to add all the actions and a `is_tracking_enabled` function. I'm not certain of the best practices when it comes to instantiating a class like this. Would a singleton make more sense?

What you have looks fine.

I've just made a few extra comments (there is a bug) and once resolved go ahead and merge into the feature branch.

File: includes/tracks/class-wc-site-tracking.php
Comment (claudiosanches) [inline includes/tracks/class-wc-site-tracking.php]: Missing docblock.

File: includes/tracks/class-wc-site-tracking.php
Comment (claudiosanches) [inline includes/tracks/class-wc-site-tracking.php]: Missing docblock.

File: includes/tracks/class-wc-site-tracking.php
Comment (claudiosanches) [inline includes/tracks/class-wc-site-tracking.php]: We don't use `@class` on WooCommerce.

File: includes/tracks/class-wc-site-tracking.php
Comment (claudiosanches) [inline includes/tracks/class-wc-site-tracking.php]: One extra line here that need to removed.

File: includes/tracks/class-wc-site-tracking.php
Comment (claudiosanches) [inline includes/tracks/class-wc-site-tracking.php]: Please move `@package` to the files docblock.

**Comment by claudiosanches** [comment]
@mikejolley @psealock I just fixed coding standards, also made changes based on Mike's feedback, and moved the includes and initialization to `ncludes/admin/class-wc-admin.php`, since this runs only in the admin.

**Comment by psealock** [comment]
Thanks for the fantastic feedback and help with this one @mikejolley @claudiosanches !

I rebased this branch against the feature branch and flagging for any last reviews.

**Comment by psealock** [comment]
>I've just made a few extra comments (there is a bug) and once resolved go ahead and merge into the feature branch.

Just re-read this part, so I'll go ahead and merge.

**Pull request merged**
